NSS-Day-Cohort-68/rare-api-rare-team-1-1
# API - Register a new user

This pull request addresses the server-side functionality of NSS-Day-Cohort-68/rare-client-rare-team-1#48.
Client-side functionality for this issue has not yet been implemented.

The purpose of this PR is to **update the create_user POST method** to:
* catch user input errors
* address incorrect naming conventions
* improve the HTTP error response messages to provide more detailed information to the client. 

**Changes made in json-server.py:**
* added try/excpet block at the top of the do_POST method to catch the JSONDecodeError that is raised when no request_body is provided. This try/except should apply to any incoming post request because each post requires something to be included in the request body.
* added the appropriate import statement from json.decoder for JSONDecodeError
* added try/except block before executing create_user to check that the required keys are provided by the client. If not, the KeyError is caught, and an informative message is returned to the client. 
* updated variable name of the create_user return value from "test" to "token", to improve code clarity.
* Updated HTTP 201 response to also return the token object to the client, rather than a success message. token contains key/value pairs for user id and valid which will may be used for authentication in the future. 

**STEPS TO TEST**
1. In the post-user branch, put a breakpoint on json-server.py at line 42 (if url['requested_resource'] = "users":
2. In postman: select POST, set the url to http://localhost:8000/users, and create a user object in the body. 
3. First, test by attempting to post an object with the expected keys: "first_name", "last_name", "username", and "email". If successful, a new user should be added to db.sqlite3 after sending the post request from postman and stepping through the code.
4. Next, test with extra keys. The same thing should happen as in step 3.
5. Next, text with a missing key (For example, delete the key/value pair, "first_name"). You should receive the error, "Incomplete user information..." as seen on line 49. 
6. Next, test with no body. You should receive the error message, "Error -- no information was provided" as seen in line 38.
